src/DotVVM.Framework/Hosting/IDotvvmRequestContext.cs

```diff
@@ -111,5 +111,7 @@ public interface IDotvvmRequestContext
         string? ResultIdFragment { get; set; }
 
         IServiceProvider Services { get; }
+        object? CommandResult { get; set; }
+        Dictionary<string, object>? CustomData { get; }
```

Comment:
When you will need to modify some object in CustomData, the AddCustomData method does not provide you the interface to do so.  
You can add some additional CustomData in static command resolution and you can also want to modify it in the OnPresenterExecutedAsync method. This is actually based on one of my use-cases.

For example, you have custom validation and you have multiple places where you just adding errors. At the end of execution, you want to filter the collection for duplicities. If there is just the add method, you cannot filter it. 

Also, CustomData was inspired by ModelState (MVC). There is used the same approach as here.

Reply:
I am fine with CustomData as a Dictionary (or IDictionary) - HttpContext.Items and other similar containers are also dictionaries. Being able to see what's in the collection can be useful.

But I agree with removing CommandResult from context. Also, I think that allowing the users to change it (in action filters) is a bad idea at all - if the users want to send something else than the result, they can use CustomData and deal with them in the postback handler.

Reply:
However, unlike ModelState and Items, this is not intended for component-componet communication.

@quigamdev I'd suggest you to rather create a request scoped service that will register your error, then do deduplication and put it into custom data. Then you'll be able to detect cases when you add an error after the deduplication.

> When you will need to modify some object in CustomData, the AddCustomData method does not provide you the interface to do so.

That's the purpose - then the order of execution can not matter. In DotVVM we generally have a big problem that things only work at certain time and this will be another source of such issues.